test/unit/cmap/handshake/client_metadata.test.ts

```diff
@@ -1,149 +1,511 @@
+            ['FUNCTION_MEMORY_MB', '1024']
+          ],
+          outcome: {
+            name: 'gcp.func',
+            memory_mb: new Int32(1024)
+          }
+        },
+        {
+          context: 'FUNCTION_REGION provided',
+          env: [
+            ['FUNCTION_NAME', 'non-empty'],
+            ['FUNCTION_REGION', 'region']
+          ],
+          outcome: {
+            name: 'gcp.func',
+            region: 'region'
+          }
+        },
+        {
+          context: 'FUNCTION_TIMEOUT_SEC provided',
+          env: [
+            ['FUNCTION_NAME', 'non-empty'],
+            ['FUNCTION_TIMEOUT_SEC', '12345']
+          ],
+          outcome: {
+            name: 'gcp.func',
+            timeout_sec: new Int32(12345)
+          }
+        }
+      ],
+      vercel: [
+        {
+          context: 'no additional metadata',
+          env: [['VERCEL', 'non-empty']],
+          outcome: {
+            name: 'vercel'
+          }
+        },
+        {
+          context: 'VERCEL_URL provided',
+          env: [
+            ['VERCEL', 'non-empty'],
+            ['VERCEL_URL', 'provided-url']
+          ],
+          outcome: {
+            name: 'vercel',
+            url: 'provided-url'
+          }
+        },
+        {
+          context: 'VERCEL_REGION provided',
+          env: [
+            ['VERCEL', 'non-empty'],
+            ['VERCEL_REGION', 'region']
+          ],
+          outcome: {
+            name: 'vercel',
+            region: 'region'
+          }
+        }
+      ]
+    };
+
+    for (const [provider, testsForEnv] of Object.entries(tests)) {
+      context(provider, () => {
+        for (const { context, env: faasVariables, outcome } of testsForEnv) {
+          const setupEnv = () => {
+            for (const [k, v] of faasVariables) {
+              if (v != null) {
+                process.env[k] = v;
+              }
+            }
+          };
+
+          const cleanupEnv = () => {
+            for (const [k] of faasVariables) {
+              delete process.env[k];
+            }
+          };
+
+          it(context, () => {
+            try {
+              setupEnv();
+              const { env } = makeClientMetadata({ driverInfo: {} });
+              expect(env).to.deep.equal(outcome);
+            } finally {
+              cleanupEnv();
+            }
+          });
+
+          it('always places name as the last key', () => {
+            try {
+              setupEnv();
+              const keys = Array.from(getFAASEnv()?.keys() ?? []);
+              expect(keys).to.have.property(`${keys.length - 1}`, 'name');
+            } finally {
+              cleanupEnv();
+            }
+          });
+        }
+      });
+    }
+
+    context('when a numeric FAAS env variable is not numerically parsable', () => {
+      before(() => {
+        process.env.AWS_EXECUTION_ENV = 'non-empty-string';
+        process.env.AWS_LAMBDA_FUNCTION_MEMORY_SIZE = '123not numeric';
```

Comment:
The leading digits are an "interesting" part of this test, if we used parseInt we would get 123, but we use `Number`/`unary +` so we can test for NaN to see if the whole string was valid.